### PR TITLE
Wei/nuget v3 support release

### PR DESCRIPTION
**Pull request by ChengTian**
@loudej  @davidfowl  @muratg  @yishaigalatzer 

Same changes as https://github.com/aspnet/dnx/pull/2058. Changing target branch to release.

### CODE REVIEW

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/HttpSource.cs [hunk 1]
```diff
@@ -131,13 +139,20 @@ internal class HttpSource
 
                 using (var response = await _client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead))
                 {
+                    if (!throwNotFound && response.StatusCode == HttpStatusCode.NotFound)
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/HttpSource.cs]: Why is this special?

Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/HttpSource.cs]: For v3 feed, FindPackagesById() becomes constructing the URI by ourselves and check whether the URI points to a JSON. A 404 page means the packages is not on this feed.

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+        internal static bool DetectNuGetV3(HttpSource httpSource, out Uri packageBaseAddress)
+        {
+            try
+            {
+                var result = httpSource.GetAsync(httpSource.BaseUri, "index_json", TimeSpan.FromHours(6)).Result;
+                using (var reader = new StreamReader(result.Stream))
+                {
+                    var content = reader.ReadToEnd();
+                    var indexJson = JObject.Parse(content);
```
Comment (JamesNK) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs]: You could load the JObject directly from the StreamReader and avoid allocating the string. You're already doing this on line 141.

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/HttpSource.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/HttpSource.cs]: Meaning you don't want to fail and retry right?

Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/HttpSource.cs]: Think about this scenario:
You have two feeds in NuGet.config: FeedA and FeedB
100 packages your app needs are only available on FeedA and 1 package your app needs is only available on FeedB.
Do we want to fail and retry the 100 packages on FeedB even they are not there?

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs (hunk 2 above)
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs]: Thanks. Fixed.

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/HttpSource.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/HttpSource.cs]: Sure, I was confirming not pushing back.

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+        public PackageInfo BuildModel(string id, string version)
+        {
+            return new PackageInfo
+            {
+                // If 'Id' element exist, use its value as accurate package Id
+                // Otherwise, use the value of 'title' if it exist
+                // Use the given Id as final fallback if all elements above don't exist
+                Id = id,
+                Version = SemanticVersion.Parse(version),
+                ContentUri = _baseUri + id.ToLowerInvariant() + "/" + version.ToLowerInvariant() + "/" + id.ToLowerInvariant() + "." + version.ToLowerInvariant() + ".nupkg",
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs]: We to ToLowerInvariant on both id and version twice.

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+            };
+        }
+
+        public async Task<Stream> OpenNuspecStreamAsync(PackageInfo package)
+        {
+            return await PackageUtilities.OpenNuspecStreamFromNupkgAsync(package, OpenNupkgStreamAsync, _reports.Information);
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs]: Why is this opening the nupkg stream?

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs (hunk 3 above)
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs]: fixing

File: src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs (hunk 4 above)
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs]: The method `OpenNuspecStreamFromNupkgAsync` will use `OpenNupkgStreamAsync` to open a nupkg stream and find out the nuspec entry and open a stream to that entry.

Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs]: That's not right for the V3 feed. The Nuspec is downloadable without getting the nupkg.

Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Restore/NuGet/NuGetv3Feed.cs]: Sorry we didn't know that. But current approach is not incorrect, it still works. Moreover,  seeking the nuspec entry in nupkg stream is faster than sending a separate nuspec download request.

**Comment by ChengTian** [comment]
Got sign-off from @davidfowl

**Pull request merged**